js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts

```diff
@@ -0,0 +1,145 @@
+        }
+      }
+    });
+  };
+
+  if (inputs[0].dims.length > 1) {
+    const shape = Object.seal<Record<typeof attributes.format, number[]>>({
+      nhwc: inputs[0].dims.slice(-1),
+      nchw: inputs[0].dims.slice(1, attributes.spatial ? 2 : undefined),
+    })[attributes.format];
+    checkShapeEqual(inputs[1].dims, shape, 'Invalid input scale');
+    checkShapeEqual(inputs[2].dims, shape, 'Invalid input B');
+    checkShapeEqual(inputs[3].dims, shape, 'Invalid input mean');
+    checkShapeEqual(inputs[4].dims, shape, 'Invalid input var');
+  } else {
+    checkShapeEqual(inputs[1].dims, [1], 'Invalid input scale');
+    checkShapeEqual(inputs[2].dims, [1], 'Invalid input B');
+    checkShapeEqual(inputs[3].dims, [1], 'Invalid input mean');
+    checkShapeEqual(inputs[4].dims, [1], 'Invalid input var');
+  }
+};
+
+const createBatchNormInferenceProgramInfo =
+    (inputs: readonly TensorView[], attributes: BatchNormAttributes): ProgramInfo => {
+      const {epsilon, spatial, format} = attributes;
+      const yShape = inputs[0].dims;
+      const outputSize = ShapeUtil.size(yShape);
+      const x = inputVariable('x', inputs[0].dataType, inputs[0].dims);
+      const scale = inputVariable('scale', inputs[1].dataType, [ShapeUtil.size(inputs[1].dims)]);
+      const bias = inputVariable('bias', inputs[2].dataType, [ShapeUtil.size(inputs[2].dims)]);
+      const inputMean = inputVariable('inputMean', inputs[3].dataType, [ShapeUtil.size(inputs[3].dims)]);
+      const inputVar = inputVariable('inputVar', inputs[4].dataType, [ShapeUtil.size(inputs[4].dims)]);
+      const y = outputVariable('y', inputs[0].dataType, yShape);
+
+      const calcCOffset = (): string => {
+        let cOffset = '';
+        if (spatial) {
+          cOffset = `let cOffset = ${
+              yShape.length === 1   ? '0u' :
+                  format === 'nhwc' ? `outputIndices[${yShape.length - 1}]` :
+                                      'outputIndices[1]'};`;
+        } else {
+          if (format === 'nchw') {
+            cOffset = `
+            ${y.indicesSet('outputIndices', '0', '0')}
+            let cOffset = ${y.indicesToOffset('outputIndices')};`;
+          } else {
+            // update C channel.
+            cOffset = `var cIndices = ${scale.type.indices}('0');
+                       cIndices[0] = outputIndices[${yShape.length - 1}];`;
+            // update D1 x ... x Dn channels.
+            for (let i = 1; i < scale.rank; i++) {
+              cOffset += `cIndices[${i}] = outputIndices[${i + 1}];`;
+            }
+            cOffset += `let cOffset = ${scale.indicesToOffset('cIndices')};`;
+          }
+        }
+        return cOffset;
+      };
+      const getInferenceModeShaderSource = (helper: ShaderHelper) => `
+  const epsilon = ${epsilon};
+  ${helper.declareVariables(x, scale, bias, inputMean, inputVar, y)}
+  ${helper.mainStart()}
+  ${helper.guardAgainstOutOfBoundsWorkgroupSizes(outputSize)}
```

Comment:
Can we use uniforms for this and any other possible/applicable variables?

Reply:
Done.